Version 1.0.0

# Introduction

In this programming assignment we will illustrate a very severe data leakage, that can often be found in competitions, where the pairs of object should be scored, e.g. predict $1$ if two objects belong to the same class and $0$ otherwise. 

The data in this assignment is taken from a real competition, and the funniest thing is that *we will not use training set at all* and achieve almost 100% accuracy score! We will just exploit the leakage.

Now go through the notebook and complete the assignment.

In [1]:
import numpy as np
import pandas as pd 
import scipy.sparse

# Load the data

Let's load the test data. Note, that we don't have any training data here, just test data. Moreover, *we will not even use any features* of test objects. All we need to solve this task is the file with the indices for the pairs, that we need to compare.

Let's load the data with test indices.

In [2]:
test = pd.read_csv('../readonly/data_leakages_data/test_pairs.csv')
test.head(10)

pairId  FirstId  SecondId
0       0     1427      8053
1       1    17044      7681
2       2    19237     20966
3       3     8005     20765
4       4    16837       599
5       5     3657     12504
6       6     2836      7582
7       7     6136      6111
8       8    23295      9817
9       9     6621      7672

For example, we can think that there is a test dataset of images, and each image is assigned a unique `Id` from $0$ to $N-1$ (N -- is the number of images). In the dataframe from above `FirstId` and `SecondId` point to these `Id`'s and define pairs, that we should compare: e.g. do both images in the pair belong to the same class or not. So, for example for the first row: if images with `Id=1427` and `Id=8053` belong to the same class, we should predict $1$, and $0$ otherwise. 

But in our case we don't really care about the images, and how exactly we compare the images (as long as comparator is binary).  

**We suggest you to try to solve the puzzle yourself first.** You need to submit a `.csv` file with columns `pairId` and `Prediction` to the grader. The number of submissions allowed is made pretty huge to let you explore the data without worries. The returned score should be very close to $1$.

**If you do not want to think much** -- scroll down and follow the instructions below.

# EDA and leakage intuition

As we already know, the key to discover data leakages is careful EDA. So let's start our work with some basic data exploration and build an intuition about the leakage.

First, check, how many different `id`s are there: concatenate `FirstId` and `SecondId` and print the number of unique elements. Also print minimum and maximum value for that vector.

In [10]:
# check how many different ids
uniqueID = pd.concat([test.FirstId, test.SecondId],axis=0).unique()
print(uniqueID)
# we can also check the frequency for each unique value in ids
fre_uniqueID = pd.concat([test.FirstId, test.SecondId],axis=0).value_counts()
print(fre_uniqueID)

[ 1427 17044 19237 ..., 25113 23636 24785]
2047     36
10989    36
6891     36
4842     36
2793     36
744      36
10477    36
8428     36
6379     36
4330     36
2281     36
232      36
1767     36
3814     36
5861     36
7908     36
9955     36
12002    36
1255     36
3302     36
5349     36
7396     36
9443     36
11490    36
743      36
2790     36
4837     36
6884     36
8931     36
10978    36
         ..
14033    21
13212    21
12510    21
19665    21
17616    21
15071    21
13022    21
12700    21
14749    21
25300    21
23251    21
21202    21
19153    21
17104    21
14559    21
21392    21
25494    21
23441    21
17298    21
19347    21
24788    21
22739    21
20690    21
18641    21
16592    21
18143    21
20190    21
22237    21
24284    21
14075    21
Length: 26325, dtype: int64


In [5]:
# print maximum and minimum in uniqueID
print(max(uniqueID)) # maximum id is 26324
print(min(uniqueID)) # minimum id is 0

26324
0


and then print how many pairs we need to classify (it is basically the number of rows in the test set)

In [6]:
# print how many rows in test dataset.
print(len(test))

368550


In [7]:
# we could also check if there are duplicated rows in test
sub_test = test[['FirstId','SecondId']]
print(sub_test.duplicated().sum()) 
# check the duplicated rows
print(sub_test.loc[sub_test.duplicated(),:])

12
        FirstId  SecondId
107647    10008     10404
123152     3505      8046
124261     5641      8046
139487      504      2799
226778     1980      8046
238033     1420      8046
239980     2627     10404
256914      332     10404
280690     6957     10404
310044     6291     10404
330339     4155     10404
353199      256      2551


we could see there are 12 rows are duplicated in test dataset.

Now print, how many distinct pairs it would be possible to create out of all "images" in the dataset?   

In [8]:
from scipy.special import comb
num_ids = len(uniqueID)
# from num_ids choose two
num_comb = comb(num_ids, 2, exact=True)
# print distinct pairs
print(num_comb)

346489650


So the number of pairs we are given to classify is very very small compared to the total number of pairs. 

To exploit the leak we need to **assume (or prove)**, that the total number of positive pairs is small, compared to the total number of pairs. For example: think about an image dataset with $1000$ classes, $N$ images per class. Then if the task was to tell whether a pair of images belongs to the same class or not, we would have $1000\frac{N(N-1)}{2}$ positive pairs, while total number of pairs was $\frac{1000N(1000N - 1)}{2}$.

Another example: in [Quora competitition](https://www.kaggle.com/c/quora-question-pairs) the task was to classify whether a pair of qustions are duplicates of each other or not. Of course, total number of question pairs is very huge, while number of duplicates (positive pairs) is much much smaller.

Finally, let's get a fraction of pairs of class `1`. We just need to submit a constant prediction "all ones" and check the returned accuracy. Create a dataframe with columns `pairId` and `Prediction`, fill it and export it to `.csv` file. Then submit to grader and examine grader's output. 

In [8]:
# predict all pairs are "1"
submit = pd.DataFrame()
submit['pairId'] = test['pairId']
submit = submit.assign(Prediction=1)
submit.to_csv('submission.csv', index=False)

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

# Building a magic feature

In this section we will build a magic feature, that will solve the problem almost perfectly. The instructions will lead you to the correct solution, but please, try to explain the purpose of the steps we do to yourself -- it is very important.

## Incidence matrix

First, we need to build an [incidence matrix](https://en.wikipedia.org/wiki/Incidence_matrix). You can think of pairs `(FirstId, SecondId)` as of edges in an undirected graph. 

The incidence matrix is a matrix of size `(maxId + 1, maxId + 1)`, where each row (column) `i` corresponds `i-th` `Id`. In this matrix we put the value `1` to the position `[i, j]`, if and only if a pair `(i, j)` or `(j, i)` is present in  a given set of pais `(FirstId, SecondId)`. All the other elements in the incidence matrix are zeros.   

**Important!** The incidence matrices are typically very very sparse (small number of non-zero values). At the same time incidence matrices are usually huge in terms of total number of elements, and it is **impossible to store them in memory in dense format**. But due to their sparsity incidence matrices **can be easily represented as sparse matrices**. If you are not familiar with sparse matrices, please see [wiki](https://en.wikipedia.org/wiki/Sparse_matrix) and [scipy.sparse reference](https://docs.scipy.org/doc/scipy/reference/sparse.html). Please, use any of `scipy.sparse` constructors to build incidence matrix. 

For example, you can use this constructor: `scipy.sparse.coo_matrix((data, (i, j)))`. We highly recommend to learn to use different `scipy.sparse` constuctors, and matrices types, but if you feel you don't want to use them, you can always build this matrix with a simple `for` loop. You will need first to create a matrix using `scipy.sparse.coo_matrix((M, N), [dtype])` with an appropriate shape `(M, N)` and then iterate through `(FirstId, SecondId)` pairs and fill corresponding elements in matrix with ones. 

**Note**, that the matrix should be symmetric and consist only of zeros and ones. It is a way to check yourself.

In [11]:
# concatenate FirstId and SecondId as row and col which are argument in coo_matrix
row_temp = test.FirstId.append(test.SecondId)
col_temp = test.SecondId.append(test.FirstId)

In [12]:
# we need to remove the repeated cases
test_temp = pd.concat([row_temp, col_temp], axis=1)
test_temp = test_temp.drop_duplicates()

In [14]:
# select the final row and col
row = test_temp[0]
col = test_temp[1]

In [15]:
# figure out data argument in coo_matrix
data = np.ones(len(row))
print(len(data))

736872


In [16]:
# create the sparse matrix
inc_mat = scipy.sparse.coo_matrix((data, (row, col)),dtype=int)

# Sanity checks
assert inc_mat.max() == 1
assert inc_mat.sum() == 736872

It is convenient to have matrix in `csr` format eventually.

In [17]:
inc_mat = inc_mat.tocsr()

## Now build the magic feature

Why did we build the incidence matrix? We can think of the rows in this matix as of representations for the objects. `i-th` row is a representation for an object with `Id = i`. Then, to measure similarity between two objects we can measure similarity between their representations. And we will see, that such representations are very good.

Now select the rows from the incidence matrix, that correspond to `test.FirstId`'s, and `test.SecondId`'s.

In [18]:
# Note, scipy goes crazy if a matrix is indexed with pandas' series. 
# So do not forget to convert `pd.series` to `np.array`
# These lines should normally run very quickly 
rows_FirstId   = inc_mat[test.FirstId.values,:]  # select rows correspond to test.FirstId's
rows_SecondId  = inc_mat[test.SecondId.values,:] # select rows correspond to test.SecondId's

Our magic feature will be the *dot product* between representations of a pair of objects. Dot product can be regarded as similarity measure -- for our non-negative representations the dot product is close to 0 when the representations are different, and is huge, when representations are similar. 

Now compute dot product between corresponding rows in `rows_FirstId` and `rows_SecondId` matrices.

In [47]:
# Note, that in order to do pointwise multiplication in scipy.sparse you need to use function `multiply`
# regular `*` corresponds to matrix-matrix multiplication

f = rows_FirstId.multiply(rows_SecondId)
print(f)
f = f.sum(axis=1)
print(f)
f = np.squeeze(np.array(f))
# Sanity check
assert f.shape == (368550, )

  (0, 541)	1
  (0, 1209)	1
  (0, 1371)	1
  (0, 1987)	1
  (0, 2934)	1
  (0, 3512)	1
  (0, 3651)	1
  (0, 3846)	1
  (0, 4474)	1
  (0, 5648)	1
  (0, 9366)	1
  (0, 11488)	1
  (0, 12360)	1
  (0, 14546)	1
  (0, 15819)	1
  (0, 17746)	1
  (0, 19083)	1
  (0, 20813)	1
  (0, 22945)	1
  (0, 24669)	1
  (1, 915)	1
  (1, 3210)	1
  (1, 4738)	1
  (1, 6872)	1
  (1, 7540)	1
  :	:
  (368548, 6285)	1
  (368548, 6951)	1
  (368548, 10002)	1
  (368548, 10398)	1
  (368548, 12994)	1
  (368548, 15181)	1
  (368548, 16456)	1
  (368548, 18384)	1
  (368548, 19719)	1
  (368548, 21449)	1
  (368548, 23578)	1
  (368549, 1206)	1
  (368549, 5028)	1
  (368549, 5429)	1
  (368549, 7832)	1
  (368549, 9171)	1
  (368549, 11270)	1
  (368549, 12140)	1
  (368549, 14326)	1
  (368549, 15599)	1
  (368549, 17526)	1
  (368549, 20593)	1
  (368549, 22323)	1
  (368549, 22724)	1
  (368549, 24454)	1
[[20]
 [14]
 [20]
 ..., 
 [14]
 [14]
 [14]]


That is it! **We've built our magic feature.** 

# From magic feature to binary predictions

But how do we convert this feature into binary predictions? We do not have a train set to learn a model, but we have a piece of information about test set: the baseline accuracy score that you got, when submitting constant. And we also have a very strong considerations about the data generative process, so probably we will be fine even without a training set. 

We may try to choose a thresold, and set the predictions to 1, if the feature value `f` is higer than the threshold, and 0 otherwise. What threshold would you choose? 

How do we find a right threshold? Let's first examine this feature: print frequencies (or counts) of each value in the feature `f`.

In [35]:
# For example use `np.unique` function, check for flags
print(np.unique(f,return_counts=True))

(array([14, 15, 19, 20, 21, 28, 35]), array([183279,    852,    546, 183799,      6,     54,     14]))


Do you see how this feature clusters the pairs? Maybe you can guess a good threshold by looking at the values? 

In fact, in other situations it can be not that obvious, but in general to pick a threshold you only need to remember the score of your baseline submission and use this information. Do you understand why and how?  

Choose a threshold below: 

In [36]:
pred = f >15# SET THRESHOLD HERE according baseline 

# Finally, let's create a submission

In [37]:
submission = test.loc[:,['pairId']]
submission['Prediction'] = pred.astype(int)

submission.to_csv('submission.csv', index=False)

Now submit it to the grader! It is not possible to submit directly from this notebook, as we need to submit a `csv` file, not a single number (limitation of Coursera platform). 

To download `submission.csv` file that you've just produced <a href='./submission.csv'>click here</a> (if the link opens in browser, right-click on it and shoose "Save link as"). Then go to [assignment page](https://www.coursera.org/learn/competitive-data-science/programming/KsASv/data-leakages/submission) and submit your `.csv` file in 'My submission' tab.


If you did everything right, the score should be very high.

**Finally:** try to explain to yourself, why the whole thing worked out. In fact, there is no magic in this feature, and the idea to use rows in the incidence matrix can be intuitively justified.

# Bonus

Interestingly, it is not the only leak in this dataset. There is another totally different way to get almost 100% accuracy. Try to find it!